### PR TITLE
Add S1121 for TS/JS ('no-nested-assignment')

### LINKED ISSUE #1408: Add S1121 for TS/JS ('no-nested-assignment')
We allow assignment in loop condition, in eslint it's possible to wrap in parentheses to ignore assignment in condition. I think we should go for this option enabled and w/o ability to configure it in SQ.

### PR DESCRIPTION
**Pull request by yassin-kammoun-sonarsource**
Fixes #1408

### Changes in the rule
The rule now raises issues for assignments in any kind of expressions except assignment expression statements, i.e. `a = 0;`.

### CODE REVIEW

File: its/ruling/src/test/expected/ts/Joust/typescript-S1121.json [hunk 1]
```diff
@@ -2,4 +2,16 @@
 'Joust:ts/components/EventLog.tsx':[
 51,
 ],
+'Joust:ts/components/GameWidget.tsx':[
```
Comment (yassin-kammoun-sonarsource) [inline its/ruling/src/test/expected/ts/Joust/typescript-S1121.json]: Should we exclude this pattern, i.e. lambda whose body is assignment expression ?

File: eslint-bridge/src/rules/no-conditional-assignment.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  );
+  const text = sourceCode.getText(node.left as estree.Node);
+  context.report({
+    message: `Extract the assignment of \"${text}\" from this expression.`,
+    loc: operator!.loc,
+  });
+}
+
+function parentOf(node: TSESTree.Node): TSESTree.Node | undefined {
+  const parent = node.parent;
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-conditional-assignment.ts]: hm, I didn't know about `node.parent`, that's something internal for estree API. Could you use classic `context.getParent()` to make sure it's always there?

File: eslint-bridge/tests/rules/no-conditional-assignment.test.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      code: `for (i = j = 0;;) {}`,
+    },
+    {
+      code: `for (i, j = 0;;) {}`,
+    },
+    {
+      code: `for (;; i = 0) {}`,
+    },
+  ],
+  invalid: [
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/no-conditional-assignment.test.ts]: you forgot to test issue location and message

File: eslint-bridge/tests/rules/no-conditional-assignment.test.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    {
+      code: `(a = 0) ? b : c;`,
+      errors: 1,
+    },
+    {
+      code: `a ? b = 0 : c;`,
+      errors: 1,
+    },
+    {
+      code: `while (a = 0) {}`,
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/no-conditional-assignment.test.ts]: we should now remove exception from rule description

File: eslint-bridge/tests/rules/no-conditional-assignment.test.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import { RuleTester } from "eslint";
+
+const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 2018 } });
+import { rule } from "../../src/rules/no-conditional-assignment";
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/no-conditional-assignment.test.ts]: I just realized that rule key should be different now as this rule has nothing to do this conditions (how about `no-nested-assignment`?)

File: eslint-bridge/src/rules/no-conditional-assignment.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import * as estree from "estree";
+import { getParent } from "eslint-plugin-sonarjs/lib/utils/nodes";
+
+export const rule: Rule.RuleModule = {
+  create(context: Rule.RuleContext) {
+    return {
+      AssignmentExpression: (node: estree.Node) => {
+        const assignment = node as estree.AssignmentExpression;
+        const ancestor = ancestorOf(assignment, context.getAncestors());
+        const parent = getParent(context);
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-conditional-assignment.ts]: tbh this logic with parents and ancestors is fuzzy. How about extracting logic checking for arrow function into a separate function to hide `parent` variable there?

Comment (yassin-kammoun-sonarsource) [reply eslint-bridge/src/rules/no-conditional-assignment.ts]: I agree, it's not nice the way it is right now.

**Pull request merged**